### PR TITLE
Custom RTL handling for ghost runs, NotoNaskhArabic test font

### PR DESCRIPTION
**Pull request by GaryQian**
Fixes https://github.com/flutter/flutter/issues/21679

Here, we track the TextDirection used for the styled run, and apply the most recent to ghost runs. This needs to be done because ICU interprets all non-language-specific (eg, normal space character) as LTR, which causes the space to be on the right side of the text in RTL languages. We handle this case manually and flip it to the opposite side.

### CODE REVIEW

File: third_party/txt/src/txt/paragraph.cc [hunk 1]
```diff
@@ -549,6 +549,19 @@ void Paragraph::Layout(double width, bool force) {
...
+    // Once the direction of a non-ghost line run has been established for the
+    // styled run, we will enforce that the ghost run use the direction of the
+    // most recently seen bidi run that is part of the styled run. This is to
+    // ensure that default-LTR whitespace gets the proper directionality when
+    // used in a RTL run.
+    std::map<size_t, TextDirection> styled_run_direction_map;
+    for (size_t i = 0; i < runs_.size(); ++i) {
+      StyledRuns::Run run = runs_.GetRun(i);
+      // Initialize TextDirection to LTR as that is what ICU initializes it to.
+      styled_run_direction_map.emplace(
```
Comment (chinmaygarde) [inline third_party/txt/src/txt/paragraph.cc]: I don't think the emplace adds much value here as the enum value is not expensive to copy.

File: third_party/txt/src/txt/paragraph.cc [hunk 2]
```diff
@@ -571,9 +589,15 @@ void Paragraph::Layout(double width, bool force) {
           line_range.end_excluding_whitespace < line_range.end &&
           bidi_run.start() <= line_range.end &&
           bidi_run.end() > line_end_index) {
+        // Use the most recent TextDirection instead of the defaulted LTR value
```
Comment (jason-simmons) [inline third_party/txt/src/txt/paragraph.cc]: `ParagraphStyle` has a field that sets the base `TextDirection` for the paragraph.  What happens in this case if you set that field to RTL?

File: third_party/txt/src/txt/paragraph.cc [hunk 3]
```diff
@@ -587,6 +598,7 @@ void Paragraph::Layout(double width, bool force) {
 
     std::vector<GlyphPosition> line_glyph_positions;
     std::vector<CodeUnitRun> line_code_unit_runs;
+    std::vector<CodeUnitRun*> ghost_code_unit_runs;
```
Comment (jason-simmons) [inline third_party/txt/src/txt/paragraph.cc]: This should no longer be necessary

File: third_party/txt/src/txt/paragraph.cc [hunk 4]
```diff
@@ -657,6 +668,8 @@ void Paragraph::Layout(double width, bool force) {
 
       layout.doLayout(text_ptr, text_start, text_count, text_size, run.is_rtl(),
                       minikin_font, minikin_paint, minikin_font_collection);
+      if (run.is_ghost() && run.is_rtl())
```
Comment (jason-simmons) [inline third_party/txt/src/txt/paragraph.cc]: Move this to after the `layout.nGlyphs() == 0` check and add a comment

**Comment by mzarsaw** [comment]
fix that with set textDirection to ltr. it's cool

**Pull request merged**